Set all 4 version compatibility options configurable

Meridiano opened the pull request:
Default values are:
- sig_scanning = false
- address_library = true
- no_struct_use = false
- layout_dependent = true

If sig_scanning is true then address_library will be forced to false.

If no_struct_use is true then layout_dependent will be forced to false.

Meridiano committed:
Set all 4 version compatibility options configurable

Default values are:
- sig_scanning = false
- address_library = true
- no_struct_use = false
- layout_dependent = true

If sig_scanning is true then address_library will be forced to false.

If no_struct_use is true then layout_dependent will be forced to false.

qudix commented on packages/c/commonlibsf/rules/plugin.lua:
```diff
@@ -72,25 +79,59 @@ rule("plugin")
+            PROJECT_VERSION_MAJOR   = semver.new(project.version() or "0.0.0"):major(),
+            PROJECT_VERSION_MINOR   = semver.new(project.version() or "0.0.0"):minor(),
+            PROJECT_VERSION_PATCH   = semver.new(project.version() or "0.0.0"):patch(),
+            OPTION_SIG_SCANNING     = tostring(configs.options.sig_scanning),
+            OPTION_ADDRESS_LIBRARY  = tostring(configs.options.address_library),
+            OPTION_NO_STRUCT_USE    = tostring(configs.options.no_struct_use),
+            OPTION_LAYOUT_DEPENDENT = tostring(configs.options.layout_dependent)
```
`tostring` shouldn't be needed here unless the config parser is failing?

Meridiano replied:
configs.options are designed to always have a correct true/false value, not nil, but I was not sure how Lua handles bool-to-string conversion. I can see you use
```
if type(result) ~= "string" then
    result = tostring(result)
end
```
next so tostring in config_map is not required indeed. If you set `OPTION_SIG_SCANNING = configs.options.sig_scanning` then `type(OPTION_SIG_SCANNING)` will be "boolean" and `OPTION_SIG_SCANNING = tostring(OPTION_SIG_SCANNING)` will be executed.

tostring in config_map is obsolete, yes.

Meridiano committed:
Remove redundant tostring conversions

Meridiano committed:
Move assert higher

If result is nil then type(result) is "nil".
"nil" ~= "string" so we execute result = tostring(nil).
After this result is a string with "nil" value and assert never cathes it.

qudix commented on packages/c/commonlibsf/rules/plugin.lua:
```diff
@@ -72,34 +79,68 @@ rule("plugin")
 
-        if configs.options == nil then
-            configs.options = {}
+        if configs.options then
+            -- options to locals
+            local signat = configs.options.sig_scanning or false
+            local addlib = configs.options.address_library == nil
+            if not addlib then
+                addlib = configs.options.address_library
+            end
+            local struct = configs.options.no_struct_use or false
+            local layout = configs.options.layout_dependent == nil
+            if not layout then
+                layout = configs.options.layout_dependent
+            end
+            -- locals to options
+            if signat then
+                configs.options.address_library = false
+            else
+                configs.options.sig_scanning = signat
+                configs.options.address_library = addlib
+            end
+            if struct then
+                configs.options.layout_dependent = false
+            else
+                configs.options.no_struct_use = struct
+                configs.options.layout_dependent = layout
+            end
+        else
+            configs.options = {
+                sig_scanning = false,
+                address_library = true,
+                no_struct_use = false,
+                layout_dependent = true
+            }
```
This can probably be simplified, but it's fine for now.

Meridiano replied:
```lua
if configs.options.sig_scanning then
    configs.options.address_library = false
else
    configs.options.sig_scanning = false
    if configs.options.address_library == nil then
        configs.options.address_library = true
    end
end
if configs.options.no_struct_use then
    configs.options.layout_dependent = false
else
    configs.options.no_struct_use = false
    if configs.options.layout_dependent == nil then
        configs.options.layout_dependent = true
    end
end
```
